### PR TITLE
8-byte long long in ACK C for i386, m68020

### PR DESCRIPTION
**Pull request by kernigh**
This branch adds the type `long long` to ACK C, which is an 8-byte integer on i386 and m68020, and causes an error on other machines. A new test set in tests/plat/long-long operates on 8-byte integers on i386 and m68020. Beware that not everything works:

- I don't provide conversions between _long long_ and floating-point types. Such a conversion might cause an error from ncg, or it might emit code that corrupts the stack.

- `446020022096LL` works, but `446020022096` gets cut to an _unsigned long_. ACK C still follows C89, not C99, for literals without the LL suffix. C89 had no _long long_. C99 would use _long long_ if the literal can't fit in _long_.

- libc has almost no support for 8-byte integers. For example, `printf("%lld", x)` doesn't work. I provide int64_t and uint64_t in <stdint.h>, but almost nothing else.

ACK languages had no 64-bit integers until now. [My mail to tack-devel](https://sourceforge.net/p/tack/mailman/message/36753108/) gave a few reasons to want 8-byte integers. Parts of the ACK assume that integers are never wider than 4 bytes, so I work around this assumption.

- Commit 054b9c8 adds the pseudo `.data8` to the assembler. `.data8` only takes a literal integer, not an expression, because expressions still use a machine-dependent integer type that might have only 4 bytes.
- Commit 1faff41 modifies ncg for i386, i80, i86, m68020, powerpc, vc4, to use `.data8` when they encounter an 8-byte constant.
- Commits 007a63d and 15950f9 add _long long_ and LL literals to the C compiler. By default, _long long_ has size -1 and causes the error, "no long long for this machine". If a platform sets _long long_ to size 8, then the C compiler emits EM code like `adi 8`. The C compiler uses a new type _writh_ (wide arithmetic) for constant operations, to avoid changing the old type _arith_.
- Most of the commits add rules like `adi 8` to i386 ncg, or tests in C; but `rol 8` and `ror 8` have tests in EM. Before these changes in 2019, the last change to mach/i386/ncg/table was in 1995.
- Commits fd27acb, e867861, 0b0c3d5 fix the m68020 assembler and add rules like `adi 8` to m68020 ncg.

EM compact assembly can't encode an 8-byte constant for _ldc_ (because the implementation of `sp_cst8` is missing), so I modified the C compiler to avoid _ldc_ with constants wider than 4 bytes. This diverts 8-byte constants to _rom_ (where EM encodes them as strings) and causes slower code. For example, the C code

```
long long x;
long long f(void) { return x | 0x100LL; }
```

becomes the i386 code

```
I_1:
.data8  256
...
mov edx,(_x)
mov ecx,(_x+4)
or edx,(I_1)
or ecx,(I_1+4)
```

instead of the simpler

```
mov edx,(_x)
mov ecx,(_x+4)
or edx,256 ! may become orb dh,1
```

To implement `sp_cst8` and enable _ldc_, one would need to widen the type _arith_ from _long_ to int64_t. This is difficult, because parts of the ACK assume that _arith_ is always _long_. This branch keeps _arith_ as _long_.

To convert between _long long_ and floating-point types, one would need to change the interface to our mach/proto/fp software floating-point, which now assumes that integers have at most 4 bytes. This also affects i386, because its 8087 library has the same interface. For m68020, I use the ack's emulator, which is missing floating-point (but a newer version of musahi might have floating-point).

### CODE REVIEW

File: lang/cem/libcc.ansi/headers/stdint.h [hunk 1]
```diff
@@ -33,16 +33,16 @@ typedef unsigned long   uint32_t;
 #endif
 #define INT32_MAX       2147483647
 #define INT32_MIN       (-2147483648)
-#define UINT32_MAX      4294967295
+#define UINT32_MAX      4294967295U
 
-/* We only get int64_t if longs are 8 bytes. */
+/* We only get int64_t if long longs are 8 bytes. */
```
Comment (davidgiven) [inline lang/cem/libcc.ansi/headers/stdint.h]: We should probably keep the _EM_LSIZE == 8 case here for LP64 architectures which use 64 bit longs and long longs.

File: mach/proto/as/comm2.y [hunk 2]
```diff
@@ -22,6 +22,7 @@ static item_t	*last_it, *o_it;
 %union {
 	word_t	y_word;
 	valu_t	y_valu;
+	int64_t	y_valu8;
```
Comment (davidgiven) [inline mach/proto/as/comm2.y]: Is it worth having two different members here, or should we just define valu_t to an int64? (I can imagine problems if something writes a y_valu8 but reads a y_valu by mistake, which wouldn't be visible on little-endian machines.)

**Comment by davidgiven** [comment]
Thank you very much for this! I left a couple of small comments but I didn't spot any obvious issues.

My only concern is that there's a lot of scope for subtle, hard-to-find breakage in cemcom.ansi due to e.g. erroneously casting a writh to an arith and losing the top 32 bits, but this kind of problem isn't new and we should go ahead anyway. I've filed #209 because we really, really should have a proper test suite for this.

File: mach/proto/as/comm2.y (hunk 2 above)
Comment (kernigh) [reply mach/proto/as/comm2.y]: valu_t gets redefined in mach/{arm, i386, i86, m68020, m68k2, mips, ns, powerpc, vax4, vc8, z8000}/as/mach0.c. Maybe we can remove those redefinitions and just let proto/as/comm0.h define valu_t as int64_t, but I haven't tried it yet.

**Comment by kernigh** [comment]
After reading your comments, I added 2 more commits to change assembler's valu_t to int64_t and add back _EM_LSIZE == 8 to <stdint.h>.

In my commit message for a434749, I mentioned startrek_c.linuxppc by mistake. It should be startrek_c.linuxmips. I'm too lazy to edit the message.

**Comment by davidgiven** [comment]
LGTM. Thank you very much!

**Pull request merged**